File: plus-f/src/main/scala/com/socialthingy/plusf/wos/ZipUtils.scala
package com.socialthingy.plusf.wos

import java.io.File
import java.nio.file.Files
import java.nio.file.StandardCopyOption.REPLACE_EXISTING
import java.util.{Collections, Optional}
import java.util.zip.{ZipEntry, ZipFile}

import scala.collection.mutable.ListBuffer
import scala.util.Try
import scala.util.control.NonFatal
import scala.collection.JavaConverters._

object ZipUtils {
  def isZipFile(file: File): Boolean = {
    val isZip = Try {
      val zf = new ZipFile(file)
      zf.close()
    }

    isZip.isSuccess
  }

  def findFiles(file: File): java.util.List[ZipEntry] = withZipFile(file)(Collections.emptyList[ZipEntry]()) { zf =>
    val entriesInFile = zf.entries()
    val matchingEntries = ListBuffer[ZipEntry]()
    while (entriesInFile.hasMoreElements) {
      val next = entriesInFile.nextElement
      val name = next.getName.toLowerCase()
      if (name.endsWith(".tap") || name.endsWith(".tzx")) {
        matchingEntries.append(next)
      }
    }
    matchingEntries.asJava
  }

  def unzipFile(zipFile: File, entry: ZipEntry): Optional[File] = withZipFile(zipFile)(Optional.empty[File]) { zf =>
    val unzipped = File.createTempFile("plusf", entry.getName)
    unzipped.deleteOnExit()
    val stream = zf.getInputStream(entry)
    try {
      Files.copy(stream, unzipped.toPath, REPLACE_EXISTING)
      Optional.of(unzipped)
    } finally {
      stream.close()
    }
  }

  private def withZipFile[T](file: File)(onError: => T)(fn: ZipFile => T): T = try {
    val zf = new ZipFile(file)
    try {
      fn(zf)
    } finally {
      zf.close()
    }
  } catch {
    case NonFatal(e) => onError
  }
}

File: plus-f/src/main/kotlin/com/socialthingy/plusf/wos/ZipUtils.kt
package com.socialthingy.plusf.wos

import java.io.File
import java.nio.file.Files
import java.nio.file.StandardCopyOption.REPLACE_EXISTING
import java.util.Optional
import java.util.zip.ZipEntry
import java.util.zip.ZipFile

object ZipUtils {
    fun isZipFile(file: File): Boolean {
        return try {
            val zf = ZipFile(file)
            zf.close()
            true
        } catch (ex: Exception) {
            false
        }
    }

    fun findFiles(file: File): List<ZipEntry> {
        return withZipFile(file, listOf(), {
            val entriesInFile = it.entries()
            val matchingEntries = mutableListOf<ZipEntry>()
            while (entriesInFile.hasMoreElements()) {
                val next = entriesInFile.nextElement()
                val name = next.name.toLowerCase()
                if (name.endsWith(".tap") || name.endsWith(".tzx")) {
                    matchingEntries.add(next)
                }
            }
            matchingEntries
        })
    }

    fun unzipFile(zipFile: File, entry: ZipEntry): Optional<File> {
        return withZipFile(zipFile, Optional.empty(), {
            val unzipped = File.createTempFile("plusf", entry.name)
            unzipped.deleteOnExit()
            val stream = it.getInputStream(entry)
            stream.use {
                Files.copy(it, unzipped.toPath(), REPLACE_EXISTING)
                Optional.of(unzipped)
            }
        })
    }

    private fun <T> withZipFile(file: File, onError: T, action: (ZipFile) -> T): T {
        val zf = ZipFile(file)
        return try {
            action(zf)
        } catch (ex: Exception) {
            onError
        } finally {
            zf.close()
        }
    }
}
